src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.EvpPKey.SlhDsa.cs

```diff
@@ -0,0 +1,152 @@
+        [LibraryImport(Libraries.CryptoNative, StringMarshalling = StringMarshalling.Utf8)]
+        private static partial SafeEvpPKeyHandle CryptoNative_SlhDsaGenerateKey(string keyType, ReadOnlySpan<byte> seed, int seedLength);
+
+        internal static SafeEvpPKeyHandle SlhDsaGenerateKey(string algorithmName, ReadOnlySpan<byte> seed)
+        {
+            SafeEvpPKeyHandle handle = CryptoNative_SlhDsaGenerateKey(algorithmName, seed, seed.Length);
+            Debug.Assert(handle != null, "handle != null");
+
+            if (handle.IsInvalid)
+            {
+                Exception ex = Interop.Crypto.CreateOpenSslCryptographicException();
+                handle.Dispose();
+                throw ex;
+            }
+
+            return handle;
+        }
+
+        // Must be kept in sync with PalKemId in native shim.
+        internal enum PalSlhDsaAlgorithmId
+        {
+            Unknown = 0,
+            SlhDsaSha2_128s = 1,
+            SlhDsaShake128s = 2,
+            SlhDsaSha2_128f = 3,
+            SlhDsaShake128f = 4,
+            SlhDsaSha2_192s = 5,
+            SlhDsaShake192s = 6,
+            SlhDsaSha2_192f = 7,
+            SlhDsaShake192f = 8,
+            SlhDsaSha2_256s = 9,
+            SlhDsaShake256s = 10,
+            SlhDsaSha2_256f = 11,
+            SlhDsaShake256f = 12,
+        }
+
+        [LibraryImport(Libraries.CryptoNative, EntryPoint = "CryptoNative_SlhDsaGetPalId")]
+        private static partial int CryptoNative_SlhDsaGetPalId(
+            SafeEvpPKeyHandle slhDsa,
+            out PalSlhDsaAlgorithmId slhDsaId);
+
+        internal static PalSlhDsaAlgorithmId GetSlhDsaAlgorithmId(SafeEvpPKeyHandle key)
+        {
+            const int Success = 1;
+            const int Fail = 0;
+            int result = CryptoNative_SlhDsaGetPalId(key, out PalSlhDsaAlgorithmId slhDsaId);
+
+            return result switch
+            {
+                Success => slhDsaId,
+                Fail => throw CreateOpenSslCryptographicException(),
+                int other => throw FailThrow(other),
+            };
+
+            static Exception FailThrow(int result)
+            {
+                Debug.Fail($"Unexpected return value {result} from {nameof(CryptoNative_SlhDsaGetPalId)}.");
+                return new CryptographicException();
+            }
+        }
+
+        [LibraryImport(Libraries.CryptoNative)]
+        private static partial int CryptoNative_SlhDsaSignPure(
+            SafeEvpPKeyHandle pkey, IntPtr extraHandle,
+            ReadOnlySpan<byte> msg, int msgLength,
+            ReadOnlySpan<byte> context, int contextLength,
+            Span<byte> destination, int destinationLength);
+
+        internal static void SlhDsaSignPure(
+            SafeEvpPKeyHandle pkey,
+            ReadOnlySpan<byte> msg,
+            ReadOnlySpan<byte> context,
+            Span<byte> destination)
+        {
+            int ret = CryptoNative_SlhDsaSignPure(
+                pkey, GetExtraHandle(pkey),
```

Comment:
I missed when this got introduced, why not use property https://github.com/dotnet/runtime/blob/main/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/SafeEvpPKeyHandle.OpenSsl.cs#L18 ? Are we using this from different assembly?

Reply:
Ah nvm, just read the comment